binary/extensions.proto

```diff
@@ -1,70 +1,77 @@
+message SimpleExtensionURI {
+    // A surrogate key used in the context of a single plan used to reference the URI associated with an extension.
+    uint32 extension_uri_surrogate_key = 1;
 
+    // The URI where this extension YAML can be retrieved. This is the "namespace" of this extension.
+    string uri = 2;
+}
 
-    message Extension {
 
-        // unique that describes a particular source for (and type of) extensions.
-        ExtensionId extension_id = 1;
+// Describes a mapping between a specific extension entity and the uri where that extension can be found.
+message SimpleExtensionDefinition {
 
-        oneof extension_type {
-            // git uri for extension types information
-            TypeExtension type_extension = 2;
-            FunctionExtension function_extension = 3;
-        }
+    oneof mapping_type {
+        ExtensionType extension_type = 1;
+        ExtensionTypeVariation extension_type_varition = 2;
+        ExtensionFunction extension_function = 3;
+    }
 
-        message TypeExtension {
-            string git_uri = 1;
-        }
+    // Describes a Type
+    message ExtensionType {
+        // references the surrogate key defined for a specific extension URI.
+        uint32 extension_uri_pointer = 1;
 
-        message FunctionExtension {
-            string git_uri = 1;
-        }
+        // A surrogate key used in the context of a single plan to reference a named type defined in that extension.
+        uint32 type_surrogate_key = 2;
 
+        // the name of the type in the defined extension YAML.
+        string name = 3;
     }
 
-    message Mapping {
-
-        oneof mapping_type {
-            TypeMapping type_mapping = 1;
-            FunctionMapping function_mapping = 2;
-        }
-
-        message TypeMapping {
-            TypeId type_id = 1;
-            ExtensionId extension_id = 2;
-            string name = 3;
-        }
-
-        message FunctionMapping {
-            FunctionId function_id = 1;
-            ExtensionId extension_id = 2;
-            string name = 3;
-            repeated Option options = 5;
-            message Option {
-                string key = 1;
-                string value = 2;
-            }
-        }
-    }
+    message ExtensionTypeVariation {
+        // references the surrogate key defined for a specific extension URI.
+        uint32 extension_uri_pointer = 1;
 
-    message ExtensionId {
-        uint32 id = 1;
-    }
+        // A surrogate key used in the context of a single plan to reference a named type variation defined in the referenced extension.
+        uint32 type_surrogate_key = 2;
 
-    message FunctionId {
-        uint64 id = 1;
+        // the name of the type in the defined extension YAML.
+        string name = 3;
     }
 
-    message TypeId {
-        uint64 id = 1;
+    message ExtensionFunction {
+        // references the surrogate key defined for a specific extension URI.
+        uint32 extension_uri_pointer = 1;
```

Comment:
Just curoius why uints?

Reply:
Just figured clearer than having negatives. No real reason beyond that. In dbs I feel like surrogate keys are usually numbered 0, 1... positive N.